waku/node/v2/message_store.nim

```diff
@@ -0,0 +1,266 @@
+  when val is openarray[byte]|seq[byte]:
+    if val.len > 0:
+      sqlite3_bind_blob(s, n.cint, unsafeAddr val[0], val.len.cint, nil)
+    else:
+      sqlite3_bind_blob(s, n.cint, nil, 0.cint, nil)
+  elif val is int32:
+    sqlite3_bind_int(s, n.cint, val)
+  elif val is uint32:
+    sqlite3_bind_int(s, int(n).cint, int(val).cint)
+  elif val is int64:
+    sqlite3_bind_int64(s, n.cint, val)
+  else:
+    {.fatal: "Please add support for the 'kek' type".}
+
+proc put*(db: MessageStore, message: WakuMessage): MessageStoreResult[void] =
+  ## Adds a message to the storage.
+  ##
+  ## **Example:**
+  ##
+  ## .. code-block::
+  ##   let res = db.put(message)
+  ##   if res.isErr:
+  ##     echo "error"
+  let s = prepare(db.env, "INSERT INTO messages (id, timestamp, contentTopic, payload) VALUES (?, ?, ?, ?);"): discard
+  checkErr bindParam(s, 1, @(message.id().data))
+  checkErr bindParam(s, 2, toUnix(now().toTime))
+  checkErr bindParam(s, 3, message.contentTopic)
+  checkErr bindParam(s, 4, message.payload)
+
+  let res =
+    if (let v = sqlite3_step(s); v != SQLITE_DONE):
+      err($sqlite3_errstr(v))
+    else:
+      ok()
+
+  # release implict transaction
+  discard sqlite3_reset(s) # same return information as step
+  discard sqlite3_clear_bindings(s) # no errors possible
+
+  res
+
+proc close*(db: MessageStore) =
+  discard sqlite3_close(db.env)
+
+  db[] = MessageStore()[]
+
+proc get*(db: MessageStore, topics: seq[ContentTopic], paging: PagingInfo, onData: DataProc): MessageStoreResult[bool] =
+  ## Retreives a message from the storage.
+  ##
+  ## **Example:**
+  ##
+  ## .. code-block::
+  ##   proc data(msg: WakuMessage) =
+  ##     echo cast[string](msg.payload)
+  ##
+  ##   let res = db.get(topics, paging, data)
+  ##   if res.isErr:
+  ##     echo "error"
+
+  var s: ptr sqlite3_stmt
+
+  let pageSize = 
+    if paging.pageSize > 0 and paging.pageSize > MaxPageSize:
+      MaxPageSize
+    else:
+      paging.pageSize 
+
+  if paging.cursor == Index():
+    let order =
+      case paging.direction
+        of FORWARD:
+          "ORDER BY timestamp ASC, id ASC"
+        of BACKWARD:
+          "ORDER BY timestamp DESC, id DESC"
+
+    var query = "SELECT timestamp, contentTopic, payload FROM messages WHERE contentTopic IN (" & join(topics, ", ") & ") " & order
+
+    if pageSize > 0:
+      query &= " LIMIT " & $pageSize
+
+    s = prepare(db.env, query): discard
+  else:
+    let order =
+      case paging.direction
+        of FORWARD:
+          "ORDER BY timestamp ASC, id ASC"
+        of BACKWARD:
+          "ORDER BY timestamp DESC, id DESC"
+
+    let where =
+      case paging.direction
+        of FORWARD:
+          "AND timestamp > ?"
+        of BACKWARD:
+          "AND timestamp < ?"
+
+    var query = "SELECT timestamp, contentTopic, payload FROM messages WHERE contentTopic IN (" & join(topics, ", ") & ") " & where & " " & order
+
+    if pageSize > 0:
+      query &= " LIMIT " & $pageSize
```

Comment:
if `pageSize == 0`, no limit is needed (`pageSize` being zero means no pagination)

Reply:
yeah, thats why we are checking if its greater than 0, otherwise we do not add the limit.